src/main/java/org/spongepowered/api/CatalogKey.java

```diff
@@ -157,14 +183,69 @@ static CatalogKey resolve(final String value) {
     @Override
     int compareTo(CatalogKey o);
 
+    /**
+     * A builder to create {@link CatalogKey}s.
+     */
     interface Builder extends ResettableBuilder<CatalogKey, Builder> {
 
+        /**
+         * Sets the namespace. Defaults to {@link CatalogKey#MINECRAFT_NAMESPACE}.
+         *
+         * @param namespace The namespace
+         * @return This builder, for chaining
+         */
         Builder namespace(String namespace);
 
+        /**
+         * Sets the namespace based on the provided {@link PluginContainer}'s
+         * identifier. Defaults to {@link CatalogKey#MINECRAFT_NAMESPACE}.
+         *
+         * @param container The namespace
+         * @return This builder, for chaining
+         */
         Builder namespace(PluginContainer container);
 
+        /**
+         * Sets the key's value.
+         *
+         * @param value The value
+         * @return This builder, for chaining
+         */
         Builder value(String value);
 
+        /**
+         * Builds the {@link CatalogKey}.
+         *
+         * @return The built catalog key
+         * @throws IllegalStateException If {@link Builder#value(String)} is not set.
+         */
         CatalogKey build() throws IllegalStateException;
     }
+
+    /**
+     * A factory to generate {@link CatalogKey}s.
+     */
+    interface Factory {
+
+        /**
+         * Resolves a catalog key from a string, using the provided default
+         * namespace if no namespace was found in {@code value}.
+         *
+         * @param value The value
+         * @param defaultNamespace The default namespace
+         * @return A new catalog key
+         */
+        CatalogKey resolve(final String value, final String defaultNamespace);
+
+        /**
+         * Resolves a catalog key from a string, using the provided
+         * {@link PluginContainer}'s identifier as the default namespace if no
+         * namespace was found in {@code value}.
+         *
+         * @param value The value
+         * @param defaultNamespace The default namespace
+         * @return A new catalog key
+         */
+        CatalogKey resolve(final String value, final PluginContainer defaultNamespace);
```

Comment:
IMO this should be default implemented or removed, as the javadocs only allow a single way of implementing it.